### PR TITLE
fix: caching account and validator in store

### PR DESCRIPTION
**Pull request by b00f**
## Description

The primary objective of this PR is to enhance the performance of the software and it also aims to reduce the syncing time. Executing each transaction requires fetching the corresponding account or validator information from LevelDB. Since database read/write operations are costly, they can significantly impact performance.

To mitigate this, we've implemented caching for accounts and validators based on their addresses. This approach allows us to write data into the DB, while reading account or validator information is performed from the cached data.

Since we are utilizing a map for caching, the complexity of fetching data is `O(1)`, which ensures optimal performance. However, if the number of accounts becomes large, for example more than 1 million, we may need to use an LRU cache.

Another modification introduced in this PR is the implementation of the clone function for validators and accounts. Within both the sandbox and store packages, we assign objects by pointer and return a cloned object.

### CODE REVIEW

**Comment by Ja7ad** [comment]
@b00f All things is okay for merging.

File: store/account_test.go [hunk 1]
```diff
@@ -12,80 +14,189 @@ import (
...
+		for i := int32(0); i < total; i++ {
+			acc, signer := account.GenerateTestAccount(i)
+			tStore.UpdateAccount(signer.Address(), acc)
+		}
+		assert.NoError(t, tStore.WriteBatch())
+		assert.Equal(t, tStore.TotalAccounts(), total)
 	})
+
+	t.Run("Get random account", func(t *testing.T) {
+		// num := util.RandInt32(total)
```
Comment (themantre) [inline store/account_test.go:132]: Update this test

File: store/store.go [hunk 2]
```diff
@@ -327,6 +332,8 @@ func (s *store) WriteBatch() error {
 	defer s.lk.Unlock()
 
 	if err := s.db.Write(s.batch, nil); err != nil {
+		// TODO: Should we panic here?
```
Comment (themantre) [inline store/store.go:335]: Let's panic here

File: types/validator/validator.go [hunk 3]
```diff
@@ -122,7 +146,7 @@ func (val *Validator) UpdateUnbondingHeight(height uint32) {
 	val.data.UnbondingHeight = height
 }
 
-// Hash return the hash of this validator.
+// Hash calculates and returns the hash of the account.
```
Comment (themantre) [inline types/validator/validator.go:149]: Fix this typo:
```
Hash calculates and returns the hash of the validator.
```

File: types/account/account.go [hunk 4]
```diff
@@ -46,33 +48,51 @@ func FromBytes(data []byte) (*Account, error) {
...
+func (acc Account) Balance() int64 {
+	return acc.data.Balance
+}
 
+// SubtractFromBalance subtracts the given amount from the account balance.
 func (acc *Account) SubtractFromBalance(amt int64) {
 	acc.data.Balance -= amt
 }
 
+// AddToBalance adds the given amount to the account balance.
```
Comment (themantre) [inline types/account/account.go:71]: Update to:
```
AddToBalance adds the given amount to the account's balance.
```

File: types/account/account.go [hunk 5]
```diff
@@ -46,33 +48,51 @@ func FromBytes(data []byte) (*Account, error) {
...
+func (acc Account) Sequence() int32 {
+	return acc.data.Sequence
+}
+
+// Stake returns the balance of the validator.
+func (acc Account) Balance() int64 {
+	return acc.data.Balance
+}
 
+// SubtractFromBalance subtracts the given amount from the account balance.
```
Comment (themantre) [inline types/account/account.go:66]: update to:
```
// SubtractFromBalance subtracts the given amount from the account's balance.
```

File: store/validator_test.go [hunk 6]
```diff
@@ -11,99 +13,190 @@ import (
...
+		assert.NoError(t, err)
+		require.NotNil(t, val)
+		assert.Equal(t, val.Number(), num)
+	})
+}
+
+func TestIterateValidators(t *testing.T) {
+	setup(t)
+
+	total := util.RandInt32(100)
```
Comment (themantre) [inline store/validator_test.go:162]: Update to:
```
total := util.RandInt32(100) + 1
```

**Comment by themantre** [comment]
Remove `TestReturnNilForNonExistingItems` test

File: store/account_test.go [hunk 7]
```diff
@@ -12,80 +14,188 @@ import (
...
 		assert.Error(t, err)
-		assert.Nil(t, v)
+		assert.Nil(t, acc)
 	})
 }
+
+func TestAccountByAddress(t *testing.T) {
+	setup(t)
+
+	total := util.RandInt32(100) + 1
```
Comment (themantre) [inline store/account_test.go:121]: No need `+1` here

File: sandbox/sandbox_test.go [hunk 8]
```diff
@@ -288,42 +288,67 @@ func TestUpdateFromOutsideTheSandbox(t *testing.T) {
...
+}
 
-	acc3 := tSandbox.accounts[addr]
-	val3 := tSandbox.validators[pub.Address()]
+func TestValidatorDeepCopy(t *testing.T) {
+	setup(t)
 
-	assert.NotEqual(t, acc1.Hash(), acc3.Account.Hash())
-	assert.NotEqual(t, val1.Hash(), val3.Validator.Hash())
+	t.Run("non existing account", func(t *testing.T) {
```
Comment (themantre) [inline sandbox/sandbox_test.go:328]: Change to:
```
non existing validator
```

**Review by themantre — APPROVED** [review]

**Pull request merged**